Auth-v2 use a single transaction per auth DML statement

nuivall opened the pull request:
Fixes https://github.com/scylladb/scylladb/issues/17738

kostja commented:
@gleb-cloudius could you please review this one too? cc @kbr-scylla

nuivall commented on auth/common.hh:
```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
+    mutations_collector(mutations_collector&) = delete;
+    mutations_collector(mutations_collector&&) = default;
+
+    // gets timestamp which should be used when building mutations
+    api::timestamp_type write_timestamp() const;
+
+    void add_mutation(mutation m);
+    void add_mutations(std::vector<mutation> ms);
+    void add_generator(generator_func f);
+    // announce commits all data, must be called only once. Nop if there was no guard provided.
+    future<> announce(::service::raft_group0_client& group0_client, seastar::abort_source& as);
```
that's because add_mutation doesn't do any splitting, as you pointed out it would break, and I don't expect it to be needed with 1-4 mutations it's being used for. 

Perhaps std::vector<> overload is a bit misleading it suggests that there can be a lot of mutations, but that's only because other internal functions return vector even despite creating 1 mutation.

piodul replied:
Right, but the current interface allows adding both generators and mutations, and mutations are appended as a generator to the end of the generator list:

```c++
    if (_muts.size() > 0) {
        _generators.emplace_back([this] (api::timestamp_type&) -> mutations_generator {
            for (auto& mut: _muts) {
                co_yield std::move(mut);
            }
            _muts.clear();
        });
    }
```

So, if one of the generators added by the user creates a lot of mutations that overflow a single group0 operation, the last generator that yields mutations added by the user will return mutations with the wrong timestamps.

Perhaps the generator list should be reversed before being merged, or the last generator should be prepended to `_generators` instead of being appended at the end.

gleb-cloudius replied:
Does generators even used in the series? I did not see it. May be we can drop it for now?

nuivall replied:
it's a draft, I have usage with generator in this PR to do. Didn't do yet as wanted to clarify the approach first, it's a lot of refactor to move everything.

nuivall commented on auth/common.hh:
```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
+// There is no guarantee about ordering of mutations.
+// If size of mutations is too big for raft too handle under single guard
+// atomicity guarantee may be violated.
+class mutations_collector {
```
yes

nuivall commented on auth/common.cc:
```diff
@@ -202,4 +203,85 @@ future<> announce_mutations(
+        _generators.clear();
+    };
+
+    co_return co_await announce_mutations_with_batching(
+            group0_client,
+            std::bind_front(&::service::raft_group0_client::start_operation, &group0_client),
+            std::move(chained_generator),
+            &as);
```
ok,responed

nuivall commented on auth/common.hh:
```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory as number of mutations may depend on user's data.
+// There is no guarantee about ordering of mutations.
+// If size of mutations is too big for raft too handle under single guard
+// atomicity guarantee may be violated.
```
yes, I did it this way as I believe it should be solved elsewhere long term (more in https://github.com/scylladb/scylladb/pull/17910#issuecomment-2013630399)

gleb-cloudius commented on auth/common.hh:
```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
+    void add_mutation(mutation m);
+    void add_mutations(std::vector<mutation> ms);
+    void add_generator(generator_func f);
+    // announce commits all data, must be called only once. Nop if there was no guard provided.
```
Outdated comment? The guard is implicit and it is an internal error if it is missing.

nuivall replied:
yes

gleb-cloudius commented on auth/common.cc:
```diff
@@ -202,4 +203,85 @@ future<> announce_mutations(
+
+    co_return co_await announce_mutations_with_batching(
+            group0_client,
+            std::bind_front(&::service::raft_group0_client::start_operation, &group0_client),
```
Don't you need to drop '_guard' since the function will try to take another one.

nuivall replied:
yes, I need to refactor this further

nuivall commented:
v2:
- added complete implementation

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8935/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8935/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:small_blue_diamond: boost/group0_test
:small_blue_diamond: boost/role_manager_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8935/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1630/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9651/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8935/)
#### Build Details:
- Duration: 6 hr 57 min
- Builder: i-0cbbed7bafe303eef (r5ad.8xlarge)

gleb-cloudius commented on service/raft/raft_group0_client.cc:
```diff
@@ -502,4 +502,100 @@ template group0_command raft_group0_client::prepare_command(write_mutations chan
+            }
+        });
+    }
+    // raft doesn't yet support streaming so we need to materialize all mutations in memory
```
What do you mean by "yet"? Raft keeps log entries in memory until truncated. There is no much opportunity for streaming.

nuivall replied:
will drop yet

gleb-cloudius commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,59 @@ public:
+    void add_generator(generator_func f);
+
+    // Commits the data, must be called only once. Nop if there was no guard provided.
+    future<> announce(::service::raft_group0_client& group0_client, std::string_view description, seastar::abort_source& as, std::optional<::service::raft_timeout> timeout);
```
If it can be called only once may be it should be marked as '&&'

nuivall replied:
I see only one potential thing if in the future we'd like to extend it to generate more than 1 transaction for whatever reason (there was something lately around auth restore)

but we can revert it then, so I'll mark  now with &&

gleb-cloudius commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,59 @@ public:
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
```
Why is it optional?

nuivall replied:
(commented in the consutructor) due to legacy v1 path which doesn't have the guard and we don't want to guard it by ifs in dozen places

gleb-cloudius commented on cql3/statements/schema_altering_statement.cc:
```diff
@@ -70,13 +71,15 @@ schema_altering_statement::execute(query_processor& qp, service::query_state& st
-        co_await grant_permissions_to_creator(state.get_client_state());
+        co_await grant_permissions_to_creator(state.get_client_state(), mc);
     }
+    co_await qp.announce_schema_statement(*this, mc);
```
But here you announce both schema changes and auth mutations through `schema_change` raft command which sends all mutation to schema merging code.

nuivall replied:
yes, any suggestions?

ptrsmrn replied:
@gleb-cloudius are we good with resolving this conversation or you have further comments?

gleb-cloudius replied:
We discussed this on a call some time ago. A rework for raft commands is needed. We should be able to send commands that include mutations for arbitrary combination of tables and trigger updates for relevant subsystems (multiple per command). It should also replace the `mixed_command` that you introduced in your "change rf" series. On that call @piodul also mentioned that auth mutations in the future  will need to trigger some work for service level subsystem so here it will not be possible since auth mutation will be sent inside a schema change command.

On that call we decided that someone need to collect the requirements and do the design. Not sure what happens after that.

ptrsmrn replied:
@gleb-cloudius OK, thanks for the background. I assume it's going to be a longer project and this PR doesn't have to be blocked until it's completed.

gleb-cloudius replied:
I do not know what happen after the call. Was there a follow up @piodul @kbr-scylla that we talked about? On one hand I understand that this is small part for this PR and it kind of works (not by design, but by chance), but on the other if everyone just keep pushing staff on top of broken design nobody will work on a fix.

kbr-scylla replied:
@kostja promised us today that this will be planned in @ptrsmrn's team for Q3.
@kostja please post a link to the issue

ptrsmrn replied:
Here's the Issue - https://github.com/scylladb/scylladb/issues/19153
I'll resolve this conversation.

gleb-cloudius commented on auth/standard_role_manager.cc:
```diff
@@ -566,8 +566,8 @@ standard_role_manager::revoke(std::string_view revokee_name, std::string_view ro
-        }).then([this, revokee_name, role_name] {
-            return this->legacy_modify_membership(revokee_name, role_name, membership_change::remove);
+        }).then([this, revokee_name, role_name, &mc] {
+            return this->modify_membership(revokee_name, role_name, membership_change::remove, mc);
```
We can probably from `this->` while we are at it.

nuivall replied:
ok

gleb-cloudius reviewed (commented):
Left comments

piodul commented on auth/standard_role_manager.cc:
```diff
@@ -546,18 +546,18 @@ standard_role_manager::grant(std::string_view grantee_name, std::string_view rol
 
-   return when_all_succeed(check_redundant(), check_cycle()).then_unpack([this, role_name, grantee_name] {
-       return this->legacy_modify_membership(grantee_name, role_name, membership_change::add);
+    return when_all_succeed(check_redundant(), check_cycle()).then_unpack([this, role_name, grantee_name, &mc] {
+        return this->modify_membership(grantee_name, role_name, membership_change::add, mc);
```
Nit: this additional indent is not necessary

nuivall replied:
I think it's fixing indent as we use 4 space (at least in this file)

piodul replied:
Ok, fine. In that case it looks like you actually missed one line after the one you fixed (`});`).

nuivall replied:
yes, since I didn't change this line I haven't spotted, will fix

nuivall commented on auth/service.hh:
```diff
@@ -174,6 +179,10 @@ public:
     }
 
+    future<> announce_mutations(::service::mutations_collector& mc) {
+        return mc.announce(_group0_client, "auth", _as, ::service::raft_timeout{});
```
same as with last PR, I am not really sure what's the use of this label, is it worth to trouble with passing around the description string?

I didn't do it because in the future it prevents unification. Currently announce_mutations is called in each statement but the follow-up can be that the whole guard is replaced with mutation_collector and then announced in a single place, in such case not possible to say which statement it is.

gleb-cloudius replied:
> same as with last PR, I am not really sure what's the use of this label, is it worth to trouble with passing around the description string?
> 

It is added to group0 history table which allows to see the entire leniarized history of all group0 operations.

nuivall replied:
that much I know, but how it's useful?

gleb-cloudius replied:
Don't you think it would be useful for debug?

nuivall replied:
I don't think burden of proof is on my side here, to prove something is usefully first prove it, not ask to prove that it is not usefull :) 

but seriously my reasoning is that we should also have some way to view mutations, no? Then if you know that it's auth and see mutations that's all you need.

nuivall replied:
fixed

piodul commented on cql3/statements/schema_altering_statement.cc:
```diff
@@ -83,6 +86,18 @@ schema_altering_statement::execute(query_processor& qp, service::query_state& st
     co_return std::move(result);
 }
 
+future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>, cql3::cql_warnings_vec>> schema_altering_statement::prepare_schema_mutations(query_processor& qp, const query_options& options, api::timestamp_type) const {
+    // derived class must implement one of prepare_schema_mutations overloads
+    on_internal_error(logger, "not implemented");
+    co_return std::make_tuple(::shared_ptr<cql_transport::event::schema_change>(nullptr), std::vector<mutation>{}, cql3::cql_warnings_vec{});
+}
+
+future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, cql3::cql_warnings_vec>> schema_altering_statement::prepare_schema_mutations(query_processor& qp, service::query_state& state, const query_options& options, service::mutations_collector& mc) const {
+    auto [ret, muts, cql_warnings] = co_await prepare_schema_mutations(qp, options, mc.write_timestamp());
+    mc.add_mutations(std::move(muts));
+    co_return std::make_tuple(ret, cql_warnings);
+}
```
By the way, regardless of whether you choose to modify all implementors of prepare_schema_mutations to take a mutations collector or you keep the current split, this looks like a good example of a change that could be moved to a separate commit.

nuivall replied:
> IMO it would be better to convert all implementors of prepare_schema_mutations

sadly it's 20 of them. but I can do it as I want to push for more usage of mutation_collector.


If all reviewers are fine with putting mutation_collector::unused() in all those places, and some additional `if` in prepare_schema_mutations when calling `mc.add_mutations` to support not calling it?

piodul replied:
> If all reviewers are fine with putting mutation_collector::unused() in all those places, and some additional `if` in `prepare_schema_mutations` when calling `mc.add_mutations` to support not calling it?

Aren't all schema altering statements supposed to use a group0 guard? You can pass a proper mutation collector with a guard in it in all cases, no?

ptrsmrn replied:
@piodul are we good with resolving this conversation or you have further comments?

nuivall commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,59 @@ public:
+    // Constructor with optional guard used to handle both legacy and current code.
+    // There is no guard for legacy code but the whole class may be passed
+    // through to simplify the flow.
+    mutations_collector(std::optional<::service::group0_guard> g) {
+        if (g) {
+            _guard = std::move(*g);
+        }
+    };
```
yes

nuivall commented on service/raft/raft_group0_client.cc:
```diff
@@ -502,4 +502,100 @@ template group0_command raft_group0_client::prepare_command(write_mutations chan
+    logger.trace("add_write_mutations_entry: {} mutations with description {}",
+            muts.size(), description);
+    auto group0_cmd = group0_client.prepare_command(
+        ::service::write_mutations{
+            .mutations{std::move(muts)},
+        },
```
not sure what you're suggesting here?

piodul replied:
Nothing in particular yet. However, if https://github.com/scylladb/scylladb/pull/18758 is merged before your changes, then you will have to use the `service_levels_change` command when applying changes to service levels, not `write_mutations`. I just wanted to point that out because the current state of the PR hardcodes the use of `write_mutations`.

nuivall replied:
let's hurry up then :D

nuivall replied:
but really we need to think it through it's not composable you if you use your specific command, do you need single transaction semantics for service level controller configuration change? Maybe you can emit your command without mutations so you achieve single push of config change but I still will be able to pack my mutations together in a separate command? cc @Jadw1

ptrsmrn replied:
@piodul are we good with resolving this conversation or you have further comments?

nuivall commented on service/raft/raft_group0_client.cc:
```diff
@@ -502,4 +502,100 @@ template group0_command raft_group0_client::prepare_command(write_mutations chan
+    return group0_client.add_entry(std::move(group0_cmd), std::move(group0_guard), as, timeout);
+}
+
+future<std::vector<mutation>> mutations_collector::materialize_mutations() {
```
ok

nuivall commented on service/raft/raft_group0_client.cc:
```diff
@@ -502,4 +502,100 @@ template group0_command raft_group0_client::prepare_command(write_mutations chan
+}
+
+future<std::pair<std::vector<mutation>, ::service::group0_guard>> mutations_collector::extract() {
+    auto muts = std::move(_muts);
+    auto muts2 = co_await materialize_mutations();
+    _muts.clear();
+    _generators.clear();
+    muts.insert(muts.end(),
+            std::make_move_iterator(muts2.begin()), std::make_move_iterator(muts2.end()));
+    co_return std::make_pair(std::move(muts), std::move(*_guard));
```
after https://github.com/scylladb/scylladb/pull/17910#discussion_r1609321147 it should be simplfied

nuivall commented on service/client_state.hh:
```diff
@@ -222,7 +222,7 @@ public:
     /// `nullptr` for internal instances.
     ///
-    const auth::service* get_auth_service() const {
+    auth::service* get_auth_service() const {
```
I can split, const is dropped if I recall  as I need to call non const method, will write more detail in splited commit message

nuivall commented on test/lib/cql_test_env.hh:
```diff
@@ -183,4 +183,6 @@ public:
 future<> do_with_cql_env(std::function<future<>(cql_test_env&)> func, cql_test_config = {}, std::optional<cql_test_init_configurables> = {});
 future<> do_with_cql_env_thread(std::function<void(cql_test_env&)> func, cql_test_config = {}, thread_attributes thread_attr = {}, std::optional<cql_test_init_configurables> = {});
 
+void do_with_mc(cql_test_env& env, std::function<void(service::mutations_collector&)> func);
```
ok

nuivall commented on auth/service.cc:
```diff
@@ -79,41 +80,61 @@ class auth_migration_listener final : public ::service::migration_listener {
-        (void)_authorizer.revoke_all(
-                auth::make_data_resource(ks_name)
-        ).handle_exception([] (std::exception_ptr e) {
+        (void)do_with(::service::mutations_collector::unused(), [this, &ks_name] (auto& mc) mutable {
```
we want to drop them in the future

piodul commented on test/boost/group0_test.cc:
```diff
@@ -237,3 +238,95 @@ SEASTAR_TEST_CASE(test_concurrent_group0_modifications) {
+                co_yield co_await insert_mut(2, 2);
+                co_yield co_await insert_mut(3, 2);
+            });
+            auto v = co_await mc.extract();
```
Hmm, ok. The semantics of `extract` and `announce` methods seem to be that they consume the object, so it is strange to me that we allow calling `announce` after `extract` - to me it sounds like it should be an error.

nuivall replied:
yes it should and will probably will be once we remove auth-v1

nuivall commented:
v3: addressed comments, notably:
- split `auth: add class for gathering mutations without immediate announce` commit into more detailed ones
- made mutations_collector::announce and extract require  r-value mc

nuivall commented:
v4: resolved conflict in service/qos/raft_service_level_distributed_data_accessor.cc

nuivall commented:
@piodul ping

nuivall commented:
v4:
- resolved conflicts with tablets alter keyspace

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9432/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9432/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:small_blue_diamond: boost/group0_test
:small_blue_diamond: boost/role_manager_test
:small_blue_diamond: boost/transport_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9432/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1866/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9889/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9432/)
#### Build Details:
- Duration: 3 hr 4 min
- Builder: spider8.cloudius-systems.com

nuivall commented on cql3/statements/detach_service_level_statement.cc:
```diff
@@ -41,11 +41,12 @@ detach_service_level_statement::execute(query_processor& qp,
-    const auto& as = *state.get_client_state().get_auth_service();
-    co_await auth::remove_attribute(as, _role_name, "service_level");
+
+    auto& as = *state.get_client_state().get_auth_service();
```
it's pre-existing code, `as` as auth_service is commonly used in this part of the codebase

nuivall commented on service/raft/raft_group0_client.cc:
```diff
@@ -503,4 +503,86 @@ template group0_command raft_group0_client::prepare_command(broadcast_table_quer
+}
+
+future<> mutations_collector::materialize_mutations() {
+    auto chained_generator = [this] (api::timestamp_type t) -> mutations_generator {
+        for (auto& generator: _generators) {
+            auto g = generator(t);
+            while (auto mut = co_await g()) {
+                co_yield std::move(*mut);
+            }
+        }
+    };
+    auto g = chained_generator(_guard->write_timestamp());
+    while (auto mut = co_await g()) {
+        _muts.push_back(std::move(*mut));
+    }
```
yes, can be simplified, leftover after some previous code

nuivall commented on auth/role_manager.hh:
```diff
@@ -172,12 +172,12 @@ public:
     /// \returns an exceptional future with nonexistant_role if the role does not exist.
     ///
-    virtual future<> set_attribute(std::string_view role_name, std::string_view attribute_name, std::string_view attribute_value) = 0;
+    virtual future<> set_attribute(std::string_view role_name, std::string_view attribute_name, std::string_view attribute_value,::service::mutations_collector& mc) = 0;
```
ok

avikivity commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,54 @@ public:
+    // Constructor with optional guard used to handle both legacy and current code.
+    // There is no guard for legacy code but the whole class may be passed
+    // through to simplify the flow.
+    mutations_collector(std::optional<::service::group0_guard> g) : _guard(std::move(g)) {}
```
Single-argument constructors should be `explicit` to prevent accidental conversions.

nuivall replied:
ok

avikivity commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,54 @@ public:
+        return mutations_collector(std::nullopt);
+    }
+
+    mutations_collector(mutations_collector&) = delete;
```
`const mutations_collector&`

nuivall replied:
ok

avikivity commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,54 @@ public:
+// Single mutations_collector object represents a single transaction.
+// If size or number of mutations is too big for raft too handle it will be
+// rejected.
+class mutations_collector {
```
Smells like it could be named group0_transaction.

nuivall replied:
in theory it's not bound to group0 specifically

avikivity replied:
It has group0 all over it.

When we develop multi-table transactions over strongly consistent tables managed by raft, we can decide whether to reuse this class, or create a new one.

kostja replied:
group0_batch perhaps or group0_mutation_batch

nuivall replied:
I can rename. But where exactly it has group0 all over it?

kostja replied:
the guard as a member?

avikivity commented on auth/standard_role_manager.cc:
```diff
@@ -480,6 +482,44 @@ standard_role_manager::modify_membership(
+    default:
+        on_internal_error(log, format("unknown membership_change value: {}", int(ch)));
+    }
+    co_await collect_mutations(_qp, mc, modify_role_members,
```
Having dynamically parsed statements rather than using prepared statements is against our own guidelines. Are they cached further down the line?

nuivall replied:
yes, in query_processor::get_mutations_internal, `auto stmt = prepare_internal(query_string);`

avikivity replied:
ok. It's not beautiful but it's not practical to try for better now.

avikivity commented on auth/common.hh:
```diff
@@ -84,14 +84,13 @@ future<> create_legacy_metadata_table_if_missing(
 // Use this function when need to perform read before write on a single guard or if
 // you have more than one mutation and potentially exceed single command size limit.
 using start_operation_func_t = std::function<future<::service::group0_guard>(abort_source*)>;
-using mutations_generator = coroutine::experimental::generator<mutation>;
```
This patch could just disappear, no? By using the other mutations_generator in the first place.

nuivall replied:
yes, it was in a single commit before, but Piotr was complaining about too big commits so I split it.

code in auth/common couldn't use code from raft_group0_client right away because it's on master already.

avikivity commented:
Looks good. I don't think mutations_collector should handle too-large transactions. We need to handle that at the root, which is commitlog, rather than adding fragile workarounds.

nuivall commented on service/raft/raft_group0_client.hh:
```diff
@@ -190,4 +191,54 @@ public:
+    void add_generator(generator_func f);
+
+    // Commits the data, nop if there was no guard provided.
+    future<> announce(::service::raft_group0_client& group0_client, std::string_view description, seastar::abort_source& as, std::optional<::service::raft_timeout> timeout) &&;
```
ok

nuivall commented:
> Looks good. I don't think mutations_collector should handle too-large transactions. We need to handle that at the root, which is commitlog, rather than adding fragile workarounds.

it's not handling it anymore, it'll simply break (throw).

avikivity commented:
> > Looks good. I don't think mutations_collector should handle too-large transactions. We need to handle that at the root, which is commitlog, rather than adding fragile workarounds.
> 
> it's not handling it anymore, it'll simply break (throw).

Ah, maybe I missed some of the conversation and some was related to previous versions. Anyway I'm happy with what's there.

nuivall commented:
v5:
- fixed auto-revoke case for drop keyspace
- addressed comments (minor things)
- move raft description to be per add_mutation(s)/add_generator call

piodul commented:
It looks like something broke in CI again:

```
 Error when executing always post condition:
18:26:38  Also:   hudson.model.Computer$TerminationRequest: Termination requested at Tue Jun 04 19:21:27 IDT 2024 by Thread[jenkins.util.Timer [#1],5,main] [id=29]
18:26:38  		at hudson.model.Computer.recordTermination(Computer.java:242)
18:26:38  		at jenkins.model.Nodes$5.run(Nodes.java:283)
18:26:38  		at hudson.model.Queue._withLock(Queue.java:1409)
```

I'll rerun the job.

nuivall commented:
v6: class rename

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9506/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9506/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:small_blue_diamond: boost/group0_test
:small_blue_diamond: boost/role_manager_test
:small_blue_diamond: boost/transport_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9506/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1921/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9946/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9506/)
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider6.cloudius-systems.com
